pkg/expr/sql/frame_db_conv.go

```diff
@@ -137,275 +149,149 @@ func convertDataType(fieldType data.FieldType) mysql.Type {
 //
 //nolint:gocyclo
 func fieldValFromRowVal(fieldType data.FieldType, val interface{}) (interface{}, error) {
-	// the input val may be nil, it also may not be a pointer even if the fieldtype is a nullable pointer type
+	// if the input interface is nil, we can return an untyped nil
 	if val == nil {
 		return nil, nil
 	}
 
 	switch fieldType {
-	// ----------------------------
-	// Int8 / Nullable Int8
-	// ----------------------------
-	case data.FieldTypeInt8:
-		v, ok := val.(int8)
-		if !ok {
-			return nil, fmt.Errorf("unexpected value type for interface %v of type %T, expected int8", val, val)
-		}
-		return v, nil
-
-	case data.FieldTypeNullableInt8:
-		vP, ok := val.(*int8)
-		if ok {
-			return vP, nil
-		}
+	case data.FieldTypeInt8, data.FieldTypeNullableInt8:
 		v, ok := val.(int8)
 		if ok {
-			return &v, nil
-		}
-		return nil, fmt.Errorf("unexpected value type for interface %v of type %T, expected int8 or *int8", val, val)
-
-	// ----------------------------
-	// Uint8 / Nullable Uint8
-	// ----------------------------
-	case data.FieldTypeUint8:
-		v, ok := val.(uint8)
-		if !ok {
-			return nil, fmt.Errorf("unexpected value type for interface %v of type %T, expected uint8", val, val)
+			if fieldType.Nullable() {
+				return &v, nil
+			}
+			return v, nil
 		}
-		return v, nil
+		return nil, fmt.Errorf("unexpected value type %v of type %T, expected int8", val, val)
 
-	case data.FieldTypeNullableUint8:
-		vP, ok := val.(*uint8)
-		if ok {
-			return vP, nil
-		}
+	case data.FieldTypeUint8, data.FieldTypeNullableUint8:
 		v, ok := val.(uint8)
 		if ok {
-			return &v, nil
-		}
-		return nil, fmt.Errorf("unexpected value type for interface %v of type %T, expected uint8 or *uint8", val, val)
-
-	// ----------------------------
-	// Int16 / Nullable Int16
-	// ----------------------------
-	case data.FieldTypeInt16:
-		v, ok := val.(int16)
-		if !ok {
-			return nil, fmt.Errorf("unexpected value type for interface %v of type %T, expected int16", val, val)
+			if fieldType.Nullable() {
+				return &v, nil
+			}
+			return v, nil
 		}
-		return v, nil
+		return nil, fmt.Errorf("unexpected value type %v of type %T, expected uint8", val, val)
 
-	case data.FieldTypeNullableInt16:
-		vP, ok := val.(*int16)
-		if ok {
-			return vP, nil
-		}
+	case data.FieldTypeInt16, data.FieldTypeNullableInt16:
 		v, ok := val.(int16)
 		if ok {
```

Comment:
This logic repeats a few times. I'm wondering if we should introduce an abstraction, and move to defining this logic once only 🤔
I don't know what that would look like, and don't feel strongly about it, just thought I'd ask the question.

Reply:
Although I moved it for two reasons:
 - One of the blocks has a two type cases in it, so error is at the bottom
 - The longer line is the error message, so that is less indented.